File: src/Broker.java
import org.zeromq.ZContext;
import org.zeromq.ZMQ;

public class Broker {

    // Loop de replicação: recebe mensagens do canal PULL (porta 5571) e as retransmite via PUB (porta 5570)
    static class ReplicationLoop implements Runnable {
        private ZContext context;

        public ReplicationLoop(ZContext context) {
            this.context = context;
        }

        @Override
        public void run() {
            // Cria o socket PULL e o socket PUB para replicação
            ZMQ.Socket repPullSocket = context.createSocket(ZMQ.PULL);
            repPullSocket.bind("tcp://*:5571");
            ZMQ.Socket repPubSocket = context.createSocket(ZMQ.PUB);
            repPubSocket.bind("tcp://*:5570");
            
            System.out.println("Broker de replicação iniciado.");
            
            try {
                while (!Thread.currentThread().isInterrupted()) {
                    // Aguarda a mensagem (recvStr retornará null se o socket for fechado)
                    String msg = repPullSocket.recvStr();
                    if (msg == null) {
                        break;
                    }
                    repPubSocket.send(msg);
                }
            } catch (Exception e) {
                // Se ocorrer alguma exceção, saia do loop
            } finally {
                repPullSocket.close();
                repPubSocket.close();
            }
        }
    }

    // Loop de notificações: recebe as notificações via PULL (porta 5581) e as retransmite via PUB (porta 5560)
    static class NotificationLoop implements Runnable {
        private ZContext context;

        public NotificationLoop(ZContext context) {
            this.context = context;
        }

        @Override
        public void run() {
            // Cria o socket PULL e o socket PUB para notificações
            ZMQ.Socket notifPull = context.createSocket(ZMQ.PULL);
            notifPull.bind("tcp://*:5581");
            ZMQ.Socket notifPub = context.createSocket(ZMQ.PUB);
            notifPub.bind("tcp://*:5560");
            
            System.out.println("Broker de notificações iniciado.");
            
            try {
                while (!Thread.currentThread().isInterrupted()) {
                    String msg = notifPull.recvStr();
                    if (msg == null) {
                        break;
                    }
                    notifPub.send(msg);
                }
            } catch (Exception e) {
                // Controle de exceção
            } finally {
                notifPull.close();
                notifPub.close();
            }
        }
    }

    public static void main(String[] args) {
        // Cria o contexto ZeroMQ que será compartilhado entre os sockets
        try (ZContext context = new ZContext()) {

            // Cria o socket frontend (ROUTER) para clientes e o backend (DEALER) para servidores
            ZMQ.Socket frontend = context.createSocket(ZMQ.ROUTER);
            frontend.bind("tcp://*:5555");

            ZMQ.Socket backend = context.createSocket(ZMQ.DEALER);
            backend.bind("tcp://*:5556");

            System.out.println("Broker iniciado e aguardando mensagens...");

            // Inicia as threads de replicação e notificações como daemon threads
            Thread replicationThread = new Thread(new ReplicationLoop(context));
            replicationThread.setDaemon(true);
            replicationThread.start();

            Thread notificationThread = new Thread(new NotificationLoop(context));
            notificationThread.setDaemon(true);
            notificationThread.start();

            // Usa o proxy para encaminhar as mensagens entre frontend e backend
            try {
                ZMQ.proxy(frontend, backend, null);
            } catch (Exception e) {
                System.out.println("Broker interrompido.");
            } finally {
                frontend.close();
                backend.close();
            }
        }
    }
}

File: src/python/broker.py
import zmq
import threading

def replication_loop(context):
    # replicação entre servidores
    rep_pull_socket = context.socket(zmq.PULL)
    rep_pull_socket.bind("tcp://*:5571")
    
    rep_pub_socket = context.socket(zmq.PUB)
    rep_pub_socket.bind("tcp://*:5570")
    
    print("Broker de replicação iniciado.")
    
    while True:
        try:
            msg = rep_pull_socket.recv_string()
            rep_pub_socket.send_string(msg)
        except zmq.ZMQError:
            break
    rep_pull_socket.close()
    rep_pub_socket.close()

def notification_loop(context):
    # notificacoes para clientes
    notif_pull = context.socket(zmq.PULL)
    notif_pull.bind("tcp://*:5581")
    
    notif_pub = context.socket(zmq.PUB)
    notif_pub.bind("tcp://*:5560")
    
    print("Broker de notificações iniciado.")
    
    while True:
        try:
            msg = notif_pull.recv_string()
            notif_pub.send_string(msg)
        except zmq.ZMQError:
            break
    notif_pull.close()
    notif_pub.close()

def main():
    context = zmq.Context()

    frontend = context.socket(zmq.ROUTER)
    frontend.bind("tcp://*:5555")

    backend = context.socket(zmq.DEALER)
    backend.bind("tcp://*:5556")

    print("Broker iniciado e aguardando mensagens...")
    
    replication_thread = threading.Thread(target=replication_loop, args=(context,), daemon=True)
    replication_thread.start()
    
    notification_thread = threading.Thread(target=notification_loop, args=(context,), daemon=True)
    notification_thread.start()
    
    try:
        zmq.proxy(frontend, backend)
    except KeyboardInterrupt:
        print("Broker interrompido.")

    frontend.close()
    backend.close()
    context.term()

if __name__ == '__main__':
    main()
